### PR TITLE
Split IDEInspectionCallbacks into CodeCompletionCallbacks and DoneParsingCallback

### PR DESCRIPTION
**Pull request by ahoppen**
Cursor info only cares about the `doneParsing` callback and not about all the `complete` functions that are now defined in `CodeCompletionCallbacks`. To make the design clearer, split `IDEInspectionCallbacks`.

rdar://105120332

### CODE REVIEW

**Review by rintaro — APPROVED** [review]
Thank you for doing this!

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: lib/Parse/ParseExpr.cpp [hunk 1]
```diff
@@ -699,8 +699,9 @@ ParserResult<Expr> Parser::parseExprKeyPath() {
         CodeCompletionExpr(pathResult.getPtrOrNull(), Tok.getLoc());
     auto *keypath = KeyPathExpr::createParsed(
         Context, backslashLoc, rootResult.getPtrOrNull(), CC, hasLeadingDot);
-    if (IDECallbacks)
-      IDECallbacks->completeExprKeyPath(keypath, DotLoc);
+    if (CodeCompletionCallbacks) {
+      CodeCompletionCallbacks->completeExprKeyPath(keypath, DotLoc);
+    }
```
Comment (compnerd) [inline lib/Parse/ParseExpr.cpp:704]:
```suggestion
    if (this->CodeCompletionCallbacks)
      this->CodeCompletionCallbacks->completeExprKeyPath(keypath, DotLoc);
```

File: lib/Parse/ParseExpr.cpp [hunk 2]
```diff
@@ -3135,8 +3138,10 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
     } else if (isArgumentList && Tok.is(tok::code_complete)) {
       // Handle call arguments specially because it may need argument labels.
       auto CCExpr = new (Context) CodeCompletionExpr(Tok.getLoc());
-      if (IDECallbacks)
-        IDECallbacks->completeCallArg(CCExpr, PreviousLoc == leftLoc);
+      if (CodeCompletionCallbacks) {
+        CodeCompletionCallbacks->completeCallArg(CCExpr,
+                                                 PreviousLoc == leftLoc);
+      }
```
Comment (compnerd) [inline lib/Parse/ParseExpr.cpp:3144]:
```suggestion
      if (this->CodeCompletionCallbacks)
        this->CodeCompletionCallbacks->completeCallArg(CCExpr,
                                                       PreviousLoc == leftLoc);
```

**Review by compnerd — COMMENTED** [review]
This seems to be a bug in the C++ parser in MSVC from VS2019.  VS2022 works properly.  There are a few issues to iron out with the migration on the CI side that I hope that we can get through once @shahmishal returns.  As a temporary measure, please accept my apologies with this workaround.

**Comment by compnerd** [comment]
@swift-ci please test

File: lib/Parse/ParseExpr.cpp [hunk 3]
```diff
@@ -734,8 +734,9 @@ ParserResult<Expr> Parser::parseExprKeyPathObjC() {
           Context, keywordLoc, lParenLoc, components, Tok.getLoc());
     }
 
-    if (IDECallbacks)
-      IDECallbacks->completeExprKeyPath(expr, DotLoc);
+    if (CodeCompletionCallbacks) {
+      CodeCompletionCallbacks->completeExprKeyPath(expr, DotLoc);
+    }
```
Comment (compnerd) [inline lib/Parse/ParseExpr.cpp:739]:
```suggestion
    if (this->CodeCompletionCallbacks)
      this->CodeCompletionCallbacks->completeExprKeyPath(expr, DotLoc);
```

File: lib/Parse/ParsePattern.cpp [hunk 4]
```diff
@@ -245,8 +245,9 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
     if (paramContext != ParameterContextKind::EnumElement) {
       auto AttrStatus = parseDeclAttributeList(param.Attrs);
       if (AttrStatus.hasCodeCompletion()) {
-        if (IDECallbacks)
-          IDECallbacks->setAttrTargetDeclKind(DeclKind::Param);
+        if (CodeCompletionCallbacks) {
+          CodeCompletionCallbacks->setAttrTargetDeclKind(DeclKind::Param);
+        }
```
Comment (compnerd) [inline lib/Parse/ParsePattern.cpp:250]:
```suggestion
        if (this->CodeCompletionCallbacks)
          this->CodeCompletionCallbacks->setAttrTargetDeclKind(DeclKind::Param);
```

File: lib/Parse/ParseDecl.cpp [hunk 5]
```diff
@@ -9375,10 +9379,11 @@ Parser::parseDeclPrecedenceGroup(ParseDeclOptions flags,
...
 
   auto checkCodeCompletion =
-      [&](IDEInspectionCallbacks::PrecedenceGroupCompletionKind SK) -> bool {
+      [&](CodeCompletionCallbacks::PrecedenceGroupCompletionKind SK) -> bool {
     if (Tok.is(tok::code_complete)) {
-      if (IDECallbacks)
-        IDECallbacks->completeInPrecedenceGroup(SK);
+      if (CodeCompletionCallbacks) {
+        CodeCompletionCallbacks->completeInPrecedenceGroup(SK);
+      }
```
Comment (compnerd) [inline lib/Parse/ParseDecl.cpp:9386]:
```suggestion
      if (this->CodeCompletionCallbacks)
        this->CodeCompletionCallbacks->completeInPrecedenceGroup(SK);
```

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Pull request merged**